src/middlewares/condom.rs

```diff
@@ -0,0 +1,1305 @@
+                        module_info.globals.len(),
+                        max_globals
+                    )
+                ));
+            }
+        }
+
+        // custom_sections: IndexMap<String, CustomSectionIndex>,
+        // Check the custom section limits
+        if let Some(max_custom_sections) = self.limits.max_custom_sections_len {
+            if module_info.custom_sections.len() > max_custom_sections {
+                return Err(MiddlewareError::new(
+                    "CondomMiddleware",
+                    "The WASM file has too many custom sections. Blocked to prevent compile-time memory bloat",
+                ));
+            }
+        }
+
+        // Check the custom section data limits
+        if let Some(max_custom_sections_data) = self.limits.max_custom_sections_data_len {
+            for (idx, data) in module_info.custom_sections_data.iter() {
+                if data.len() > max_custom_sections_data {
+                    let section_name = module_info
+                        .custom_sections
+                        .get_index(idx.as_u32() as usize)
+                        .map(|(name, _)| name.to_owned())
+                        .unwrap_or("Unknown section".to_string());
+
+                    return Err(MiddlewareError::new(
+                        "CondomMiddleware",
+                        format!("The WASM file custom section {:?} named '{}' with size {} is too big (limit: {}). Blocked to prevent compile-time memory bloat",
+                            idx,
+                            section_name,
+                            data.len(),
+                            max_custom_sections_data,
+                        ),
+                    ));
+                };
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Generates a `FunctionMiddleware` for a given function.
+    fn generate_function_middleware(&self, _: LocalFunctionIndex) -> Box<dyn FunctionMiddleware> {
+        // This middleware doesn't need to modify functions, so we return a no-op middleware.
+        Box::new(NoOpFunctionMiddleware)
+    }
+}
+
+/// A no-op function middleware used as a placeholder.
+struct NoOpFunctionMiddleware;
+
+impl fmt::Debug for NoOpFunctionMiddleware {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("NoOpFunctionMiddleware").finish()
+    }
+}
+
+impl FunctionMiddleware for NoOpFunctionMiddleware {
+    fn feed<'a>(
+        &mut self,
+        operator: wasmer::wasmparser::Operator<'a>,
+        state: &mut MiddlewareReaderState<'a>,
+    ) -> Result<(), MiddlewareError> {
+        state.push_operator(operator);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::sync::Arc;
+    use wasmer::{sys::EngineBuilder, wat2wasm, CompilerConfig, Cranelift, Module, Store};
+
+    const MEMORY_LIMIT_1GB: u64 = 1024 * 1024 * 1024;
```

Comment:
How was this limit chosen? Do we need stricter limits?
It is only the size of the memory taken by the compiler, and not the size of the compiled module that will be cached I assume.

Reply:
First, please note that this is only used for tests.

It prevents tests that are generating wasm file with an increasing number of items to saturate the ram of the machine they are run on.
It was more useful during the early testing phase in order to determine relative weight of various kind on items (function, global variables…) but I choose to keep it because I don't want my machine to be stuck again (I guess nobody wants :) ).